### PR TITLE
feat: TLS passthrough

### PR DESCRIPTION
**Pull request by wyattrees**
Adds configuration for passing encrypted requests through HAProxy, to MAAS. 

Requires merging of https://github.com/canonical/maas-charms/pull/220. Until it is merged, you can use `latest/edge/passthrough-test` for the maas-region charm channel

### CODE REVIEW

**Comment by wyattrees** [comment]
Note: the git history is a bit messed up on this, since my branch was originally branched off of my TLS Termination branch

File: anvil-python/anvil/commands/haproxy.py [hunk 1]
```diff
@@ -42,12 +42,17 @@
...
 )
-LOG = logging.getLogger(__name__)
+VALID_TLS_MODES = ["termination", "passthrough"]
 
 
-def validate_cert_file(filepath: str) -> None:
-    if filepath == "":
-        return
+def validate_cert_file(filepath: str | None) -> None:
+    if filepath is None:
```
Comment (skatsaounis) [inline anvil-python/anvil/commands/haproxy.py:49]: Why do we need to check on None rather than empty string? I am asking since `None` is bringing up the prompt when trying to apply a manifest that does not want to set ssl cert and path. Including the default one. Which is wrong behavior and that's why the change to `""` was chosen as part of #59

File: anvil-python/anvil/commands/haproxy.py [hunk 2]
```diff
@@ -58,9 +63,13 @@ def validate_cert_file(filepath: str) -> None:
         raise ValueError(f"Permission denied when trying to read {filepath}")
 
 
-def validate_key_file(filepath: str) -> None:
-    if filepath == "":
-        return
+def validate_key_file(filepath: str | None) -> None:
+    if filepath is None:
```
Comment (skatsaounis) [inline anvil-python/anvil/commands/haproxy.py:67]: Same as the previous comment

File: anvil-python/anvil/commands/haproxy.py [hunk 3]
```diff
@@ -90,14 +103,19 @@ def haproxy_questions() -> dict[str, questions.PromptQuestion]:
         ),
         "ssl_cert": questions.PromptQuestion(
             "Path to SSL Certificate for HAProxy (enter nothing to skip TLS)",
-            default_value="",
+            default_value=None,
```
Comment (skatsaounis) [inline anvil-python/anvil/commands/haproxy.py:106]: part of the previous comments

File: anvil-python/anvil/commands/haproxy.py [hunk 4]
```diff
@@ -90,14 +103,19 @@ def haproxy_questions() -> dict[str, questions.PromptQuestion]:
...
             validation_function=validate_cert_file,
         ),
         "ssl_key": questions.PromptQuestion(
             "Path to private key for the SSL certificate (enter nothing to skip TLS)",
-            default_value="",
+            default_value=None,
```
Comment (skatsaounis) [inline anvil-python/anvil/commands/haproxy.py:111]: part of the previous comments

File: anvil-python/anvil/commands/haproxy.py [hunk 5]
```diff
@@ -90,14 +103,19 @@ def haproxy_questions() -> dict[str, questions.PromptQuestion]:
...
             validation_function=validate_key_file,
         ),
+        "tls_mode": questions.PromptQuestion(
+            'TLS termination at HA Proxy ("termination"), passthrough to MAAS ("passthrough"), or no TLS ("")?',
+            default_value="",
```
Comment (skatsaounis) [inline anvil-python/anvil/commands/haproxy.py:116]: I would prefer this as it is more natural:

```suggestion
            default_value="disabled",
```

File: anvil-python/anvil/commands/haproxy.py [hunk 6]
```diff
@@ -152,13 +170,15 @@ def has_prompts(self) -> bool:
...
-        variables.setdefault("ssl_key", "")
+        variables.setdefault("ssl_cert_content", "")
+        variables.setdefault("ssl_key_content", "")
+        variables.setdefault("tls_mode", "")
 
         # Set defaults
         self.preseed.setdefault("virtual_ip", "")
-        self.preseed.setdefault("ssl_cert", "")
-        self.preseed.setdefault("ssl_key", "")
+        self.preseed.setdefault("ssl_cert", None)
```
Comment (skatsaounis) [inline anvil-python/anvil/commands/haproxy.py:179]: part of previous comments

File: anvil-python/anvil/commands/haproxy.py [hunk 7]
```diff
@@ -152,13 +170,15 @@ def has_prompts(self) -> bool:
...
+        self.preseed.setdefault("ssl_key", None)
```
Comment (skatsaounis) [inline anvil-python/anvil/commands/haproxy.py:180]: part of previous comments

**Review by skatsaounis — CHANGES_REQUESTED** [review]

File: anvil-python/anvil/commands/haproxy.py (hunk 1 above)
Comment (wyattrees) [reply anvil-python/anvil/commands/haproxy.py]: I think I got confused while rebasing and accidentally reverted this back to checking `None`. I've changed it back now.

File: anvil-python/anvil/commands/haproxy.py [hunk 8]
```diff
@@ -98,6 +112,11 @@ def haproxy_questions() -> dict[str, questions.PromptQuestion]:
             default_value="",
             validation_function=validate_key_file,
         ),
+        "tls_mode": questions.PromptQuestion(
+            'TLS termination at HA Proxy ("termination"), passthrough to MAAS ("passthrough"), or no TLS ("")?',
+            default_value="",
+            validation_function=validate_tls_mode,
+        ),
```
Comment (skatsaounis) [inline anvil-python/anvil/commands/haproxy.py:119]: we should also add a `ssl_ca_cert` in case of pass through for self signed certificates

File: anvil-python/anvil/commands/haproxy.py [hunk 9]
```diff
@@ -183,21 +209,16 @@ def extra_tfvars(self) -> dict[str, Any]:
...
             variables["haproxy_port"] = 443
-            variables["haproxy_services_yaml"] = self.get_tls_services_yaml()
+            variables["haproxy_services_yaml"] = self.get_tls_services_yaml(
+                variables["tls_mode"]
+            )
         else:
             variables["haproxy_port"] = 80
 
         # Terraform does not need the content of these answers
-        variables.pop("ssl_cert", None)
```
Comment (skatsaounis) [inline anvil-python/anvil/commands/haproxy.py:199]: Why are those two lines removed?

File: anvil-python/anvil/commands/haproxy.py [hunk 10]
```diff
@@ -183,21 +209,16 @@ def extra_tfvars(self) -> dict[str, Any]:
             self.client, self._HAPROXY_CONFIG
         )
 
-        cert_filepath = variables["ssl_cert"]
-        key_filepath = variables["ssl_key"]
-        if cert_filepath != "" and key_filepath != "":
-            with open(cert_filepath) as cert_file:
-                variables["ssl_cert_content"] = cert_file.read()
```
Comment (skatsaounis) [inline anvil-python/anvil/commands/haproxy.py:190]: we want the file content to pass only to terraform and keep the user's answer on the answer dict. Why are those lines removed and the lines above added?

File: cloud/etc/deploy-maas-region/variables.tf [hunk 11]
```diff
@@ -49,7 +49,19 @@ variable "enable_haproxy" {
...
+  default     = ""
+}
+
+variable "ssl_cert_content" {
+  description = "SSL certificate for tls_mode=passthrough"
+  type        = string
+  default     = ""
+}
+
+variable "ssl_key_content" {
```
Comment (skatsaounis) [inline cloud/etc/deploy-maas-region/variables.tf:63]: `ssl_cac_cert` is missing from here too, should we add it.

File: anvil-python/anvil/commands/haproxy.py (hunk 10 above)
Comment (skatsaounis) [reply anvil-python/anvil/commands/haproxy.py:190]: In addition, This is the place to check that both of them are set and if not fail. The code that is calling this function will exit if TerraformException is raised. You can raise it yourself:

```py
        except TerraformException as e:
            return Result(ResultType.FAILED, str(e))
```

https://github.com/canonical/snap-openstack/blob/anvil-0.1/sunbeam-python/sunbeam/jobs/steps.py#L97-L112

File: anvil-python/anvil/commands/maas_region.py [hunk 12]
```diff
@@ -76,8 +76,10 @@ def extra_tfvars(self) -> dict[str, Any]:
         haproxy_vars: dict[str, Any] = questions.load_answers(
             self.client, HAPROXY_CONFIG_KEY
         )
-        if enable_haproxy and haproxy_vars.get("ssl_cert", "") != "":
-            variables["tls_mode"] = "termination"
+        variables["tls_mode"] = haproxy_vars["tls_mode"]
+        if variables["tls_mode"] == "passthrough":
+            variables["ssl_cert_content"] = haproxy_vars["ssl_cert_content"]
```
[the remark this answers is not in the document]
Comment (skatsaounis) [reply anvil-python/anvil/commands/maas_region.py:81]: In addition, an error can be raised too, like explained earlier.

File: anvil-python/anvil/commands/haproxy.py [hunk 13]
```diff
@@ -98,6 +112,11 @@ def haproxy_questions() -> dict[str, questions.PromptQuestion]:
             default_value="",
             validation_function=validate_key_file,
         ),
```
Comment (skatsaounis) [inline anvil-python/anvil/commands/haproxy.py:114]: Since we are going to use the ssl params for both haproxy and maas-region from now on, we should update the entire logic to have a common set ot ssl_questions, displayed in an `ssl` section. In addition, what happens if we have nodes with region role with passthrough but no node with role haproxy? We need to add a prompt to maas-region too. Both of the prompts, whichever asked first should get the answers user provided (if not manifest) and the second one should avoid prompting if someone has already provided the answers.

**Review by skatsaounis — CHANGES_REQUESTED** [review]

File: anvil-python/anvil/commands/haproxy.py [hunk 14]
```diff
@@ -42,12 +42,17 @@
...
     1200  # 15 minutes, adding / removing units can take a long time
 )
-LOG = logging.getLogger(__name__)
+VALID_TLS_MODES = ["termination", "passthrough", "disabled"]
 
 
 def validate_cert_file(filepath: str) -> None:
     if filepath == "":
-        return
+        # This question is only asked when tls_mode is "termination" or "passthrough"
```
[the remark this answers is not in the document]
Comment (wyattrees) [reply anvil-python/anvil/commands/haproxy.py:50]: ok, I'm going to add some checks in the `prompt` method then, since they really do need to provide a cert/key pair when enabling TLS

[the remark this answers is not in the document]
Comment (wyattrees) [reply anvil-python/anvil/commands/haproxy.py:50]: nevermind about the checks in `prompt`, I see your comment below now.

File: anvil-python/anvil/commands/haproxy.py (hunk 9 above)
Comment (wyattrees) [reply anvil-python/anvil/commands/haproxy.py:199]: Because I wasn't saving them initially. Now that I'm reverting that change, I'll add these back.

File: anvil-python/anvil/commands/haproxy.py [hunk 15]
```diff
@@ -71,6 +71,19 @@ def validate_key_file(filepath: str) -> None:
...
+    if filepath == "":
+        return
+    if not os.path.isfile(filepath):
+        raise ValueError(f"{filepath} does not exist")
+    try:
+        # just make sure we can open the file
+        with open(filepath):
+            pass
+    except PermissionError:
+        raise ValueError(f"Permission denied when trying to read {filepath}")
```
Comment (SK1Y101) [inline anvil-python/anvil/commands/haproxy.py:84]: You could combine the file check with the try/except

```python3
try:
    with open(filepath):
        pass
except FileNotFoundError:
    raise ValueError(f"{filepath} does not exist")
except PermissionError:
    raise ValueErrir(f"Permission denied when trying to read {filepath}")
```

**Review by SK1Y101 — DISMISSED** [review]
Not much extra to add, looks good

File: anvil-python/anvil/commands/haproxy.py [hunk 16]
```diff
@@ -154,24 +189,33 @@ def prompt(self, console: Console | None = None) -> None:
...
+        self.preseed.setdefault("ssl_cacert", "")
+        self.preseed.setdefault("tls_mode", "disabled")
 
+        qs = haproxy_questions()
+        qs.update(tls_questions(HAPROXY_VALID_TLS_MODES))
         haproxy_config_bank = questions.QuestionBank(
-            questions=haproxy_questions(),
+            questions=qs,
             console=console,
             preseed=self.preseed.get("haproxy"),
```
Comment (skatsaounis) [inline anvil-python/anvil/commands/haproxy.py:207]: Your plan is to consume the defaults (preseed) from section `haproxy`. This is fine but you need to put inside preesed both haproxy_questions and also tls_questions. Or you can have two sections. Please take a look at `anvil-python/anvil/provider/local/deployment.py#85`. Whatever you decide, please make sure that the preseed (and manifest) include all fields into one or two sections.

File: anvil-python/anvil/commands/maas_region.py [hunk 17]
```diff
@@ -62,22 +73,82 @@ def __init__(
...
+
+        self.preseed.setdefault("ssl_cert", "")
+        self.preseed.setdefault("ssl_key", "")
+        self.preseed.setdefault("ssl_cacert", "")
+        self.preseed.setdefault("tls_mode", "disabled")
+
+        maas_region_config_bank = questions.QuestionBank(
+            questions=tls_questions(MAAS_REGION_VALID_TLS_MODES),
+            console=console,
+            preseed=self.preseed.get("maas-region"),
```
Comment (skatsaounis) [inline anvil-python/anvil/commands/maas_region.py:109]: This one is problematic for the same reason as getting haproxy from preseed. At the moment, there is no `maas-region` section in the preseed (and manifest)

File: anvil-python/anvil/commands/haproxy.py [hunk 18]
```diff
@@ -183,21 +227,30 @@ def extra_tfvars(self) -> dict[str, Any]:
...
+            with open(variables["ssl_cert"]) as cert_file:
                 variables["ssl_cert_content"] = cert_file.read()
-            with open(key_filepath) as key_file:
+            with open(variables["ssl_key"]) as key_file:
                 variables["ssl_key_content"] = key_file.read()
-            variables["haproxy_port"] = 443
-            variables["haproxy_services_yaml"] = self.get_tls_services_yaml()
+            if variables["ssl_cacert"]:
+                with open(variables["ssl_cacert"]) as cacert_file:
+                    variables["ssl_cacert_content"] = cacert_file.read()
```
Comment (skatsaounis) [inline anvil-python/anvil/commands/haproxy.py:245]: I am curious how haproxy is going to use cacert. Unless it is not using it at all. Could you please explain why this is not used only by maas-region?

**Review by skatsaounis — CHANGES_REQUESTED** [review]
Thank you for taking care of the comments. I added a couple more but hopefully the last ones. In addition to those, could you also rebase to latest main?

File: anvil-python/anvil/commands/haproxy.py [hunk 19]
```diff
@@ -42,31 +41,44 @@
...
+    except PermissionError:
+        raise ValueError(f"Permission denied when trying to read {filepath}")
+
+
+def validate_cacert_chain(filepath: str) -> None:
+    if filepath == "":
+        return
+    try:
+        # just make sure we can open the file
+        with open(filepath):
```
[the remark this answers is not in the document]
Comment (wyattrees) [reply anvil-python/anvil/commands/haproxy.py:78]: I'm a little confused between this comment and the one above--are we leaving the check or removing it? The purpose behind checking was more to make sure they didn't accidentally pass a private key file as a cert or vice versa, not to fully check that the certificate is actually a valid one. I'm going to leave the check as the suggestion above gives for now.

File: anvil-python/anvil/commands/haproxy.py (hunk 18 above)
Comment (wyattrees) [reply anvil-python/anvil/commands/haproxy.py:245]: This is here for if the user selects `passthrough` as the `tls_mode`. All of the cert/key/cacert get passed on to `maas-region` if this is the case. So, haproxy doesn't ever use the cacert, but since we're asking TLS questions here when HA proxy is enabled, I figured it's best that all of the TLS questions come at the same time, even if the answers are intended to be used by `maas-region`

Comment (wyattrees) [reply anvil-python/anvil/commands/haproxy.py:245]: oh, I now see the problem here--I think this logic can be removed from `extra_tfvars` here

File: cloud/etc/deploy-maas-region/variables.tf [hunk 20]
```diff
@@ -49,7 +49,25 @@ variable "enable_haproxy" {
 }
 
 variable "tls_mode" {
-  description = "TLS Mode for MAAS Region charm ('', 'termination', or 'passthrough')"
+  description = "TLS Mode for MAAS Region charm ('disabled', 'termination', or 'passthrough')"
+  type        = string
+  default     = ""
```
Comment (skatsaounis) [inline cloud/etc/deploy-maas-region/variables.tf:54]:
```suggestion
  default     = "disabled"
```

File: anvil-python/anvil/commands/haproxy.py [hunk 21]
```diff
@@ -81,23 +93,46 @@ def validate_virtual_ip(value: str) -> None:
...
             default_value="",
             validation_function=validate_key_file,
         ),
+        "ssl_cacert": questions.PromptQuestion(
+            "Path to CA cert chain, for use with self-signed SSL certificates (enter nothing to skip)",
+            default_value="",
+            validation_function=validate_cacert_chain,
+        ),
+        "tls_mode": questions.PromptQuestion(
+            f"TLS mode: {tls_modes}?",
```
Comment (skatsaounis) [inline anvil-python/anvil/commands/haproxy.py:122]:
```suggestion
            f"TLS mode: {tls_modes}",
```

**Review by skatsaounis — APPROVED** [review]
LGTM +1
